### DISCUSSION TITLE
When is sitemap() invoked and best practices?

### DISCUSSION THREAD

**Discussion by ben519** [vercel/next.js#62029] [App Router]
I'm reading the docs for [sitemap](https://nextjs.org/docs/app/api-reference/file-conventions/metadata/sitemap). It states

> You can use the `sitemap.(js|ts)` file convention to programmatically generate a sitemap by exporting a default function that returns an array of URLs.

However, it doesn't say anything about when this function is invoked. I assume it's invoked at build time, but this raises some issues / questions:

1. What if I don't rebuild my site for weeks? The sitemap will become outdated as my site's users add new pages. What's the best way to overcome this issue?
2. Presumably, every time I build my site, `sitemap()` is invoked. In my case, it'll fetch every record in a large database table of _posts_. This could become costly if I have to rebuild my site multiple times in one day.

Any advice on this front would be appreciated!

**Comment by spirituelle** [comment]
Use cach system to revalidate your data when needed.
your request should look like this : 
` const res = await fetch(${process.env.BACKEND_URL}/api/site-map, { next: { tags: [ 'sitemap'] }});`
and then you have to set up a revalidation system under /app/api/revalidate/route.ts, it should look like this
`import { revalidateTag } from 'next/cache'`
`import { NextRequest, NextResponse } from 'next/server'`

`export async function POST(request: NextRequest) {`
 ` const secret = request.nextUrl.searchParams.get('secret')`
 ` const tag = request.nextUrl.searchParams.get('tag')`
 
 ` if (secret !== process.env.REVALIDATION_SECRET_TOKEN) {`
 `   return NextResponse.json({ message: 'Invalid secret' }, { status: 401 })`
 ` }`
 
 ` if (!tag) {`
  `  return NextResponse.json({ message: 'Missing tag param' }, { status: 400 })`
 ` }`
 
`  revalidateTag(tag)`
 
 ` return NextResponse.json({ revalidated: true, now: Date.now() })`
then whenever there new data or once a day or whatever you can send a POST request to yourwebsite/api/revalidate with the tag you want to revalidate (in this case sitemap) and the secret.
i hope this helpfull. it work fine for me
